packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,285 @@
+    _previousFatalExceptionHandler(initialError);
+  }
+
+  NSString *name;
+  NSString *reason;
+  NSMutableDictionary *userInfo;
+  if ([initialError isKindOfClass:[NSError class]]) {
+    // format these keys similar to RN -- RCTFatal in RCTAssert.m
+    name = [NSString stringWithFormat:@"%@: %@", RCTFatalExceptionName, ((NSError *)initialError).localizedDescription];
+    reason = RCTFormatError(((NSError *)initialError).localizedDescription, ((NSError *)initialError).userInfo[RCTJSStackTraceKey], 175);
+    userInfo = [((NSError *)initialError).userInfo mutableCopy];
+    userInfo[@"RCTUntruncatedMessageKey"] = RCTFormatError(((NSError *)initialError).localizedDescription, ((NSError *)initialError).userInfo[RCTJSStackTraceKey], -1);
+  } else if ([initialError isKindOfClass:[NSException class]]) {
+    name = ((NSException *)initialError).name;
+    reason = ((NSException *)initialError).reason;
+    userInfo = ((NSException *)initialError).userInfo.mutableCopy;
+  } else {
+    NSAssert(NO, @"Shouldn't add object types other than NSError or NSException to _encounteredErrors");
+  }
+
+  userInfo[@"EXUpdatesLaterEncounteredErrors"] = [_encounteredErrors copy];
+  [_delegate throwException:[NSException exceptionWithName:name reason:reason userInfo:userInfo]];
+}
+
+# pragma mark - monitoring / lifecycle
+
+- (void)_registerObservers
+{
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_handleJavaScriptDidFailToLoad) name:RCTJavaScriptDidFailToLoadNotification object:nil];
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_handleContentDidAppear) name:RCTContentDidAppearNotification object:nil];
+}
+
+- (void)_unregisterObservers
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self name:RCTJavaScriptDidFailToLoadNotification object:nil];
+  [[NSNotificationCenter defaultCenter] removeObserver:self name:RCTContentDidAppearNotification object:nil];
+}
+
+- (void)_handleJavaScriptDidFailToLoad
+{
+  [self _unregisterObservers];
+}
+
+- (void)_handleContentDidAppear
+{
+  [self _unregisterObservers];
+  [_delegate markSuccessfulLaunchForLaunchedUpdate];
+  dispatch_async(_errorRecoveryQueue, ^{
+    self->_rctContentHasAppeared = YES;
+    // the launch now counts as "successful" so we don't want to roll back;
+    // remove any extraneous tasks from the pipeline as such
+    self->_pipeline = [self->_pipeline objectsAtIndexes:[self->_pipeline indexesOfObjectsPassingTest:^BOOL(NSNumber *obj, NSUInteger idx, BOOL *stop) {
+      return obj.integerValue == EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate || obj.integerValue == EXUpdatesErrorRecoveryTaskCrash;
+    }]].mutableCopy;
+  });
+  // wait 10s before unsetting error handlers; even though we won't try to
+  // relaunch if our handlers are triggered after now, we still want to give
+  // the EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate task a reasonable
+  // amount of time to run and fetch a new update, in case there is one
```

Comment:
If I understand correctly, the intent of this code is to wait after the app launches successfully by our definition (initial screen was rendered) so if there's an error in the next 10 seconds, we'll begin checking for + loading an update. Loading the update is given `_remoteLoadTimeout` milliseconds to run.

```suggestion
  // wait 10s before unsetting error handlers; even though we won't try to
  // relaunch if our handlers are triggered after now, we still want to give
  // the EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate task a reasonable
  // window of time to start and check for a new update is there is one
```